Port to Vite and Typescript

Saghen opened the pull request:
I recently discovered this fantastic project and, while getting the extension running on Firefox, I noticed the extension could use a fresh coat of paint. In general, I've attempted to improve the maintainability of the codebase through refactoring and the switch to Typescript while attempting to avoid complexity from a bundler.

- Switch to Vite and Typescript with auto-reload development environment
- Use ENVs for detection of testing and consent 
- Support both Manifest V2 and V3 for Firefox and Chrome respectively
- Use package.json for all scripts
- Remove dependency on ua-parser-js
- Separate all calls to `browser.storage.local` into `storage.ts`
- Misc refactoring of background, popup and consent

Currently, I'm using a fork of `aw-client-js` with support for `fetch`. Axios uses XHR behind the scenes which has been removed in Service Workers, a requirement for Manifest V3. Fetch is only supported on Node 18+ by default so I'll need to revisit those changes. https://github.com/Saghen/aw-client-js

ErikBjare commented:
Wow, awesome work. I'm a bit daunted by such a massive overhaul PR, but a few initial thoughts:

 - Keep the `media` submodule, it's there for a reason (to keep all assets in one place and avoid having img blobs in several repos).
 - Will the `ua-parser-js` deprecation/removal impact browser detection? (looks okay, but not sure)
 - Will Vite generate a non-minified, human-readable, output file? (Mozilla/Chrome addon review requires it)
 - What's the "{{chrome/firefox}}.manifest..." magic?
 - To what extend have you tested it? :)

This PR will probably supersede #100

Saghen commented:
> Will the ua-parser-js deprecation/removal impact browser detection?
Tested on Firefox, Chrome and Opera but haven't tested on Safari

> What's the "{{chrome/firefox}}.manifest..." magic?
Docs can be found here: https://vite-plugin-web-extension.aklinker1.io/guide/configuration.html#browser-specific-manifest-fields

> To what extend have you tested it? :)
Tested the consent and popup on both Chrome and Firefox. Using it in a test environment alongside the production extension to see if there's any discrepancies but so far so good

Saghen commented:
Heads up that I haven't abandoned this! Waiting on the axios support for fetch https://github.com/axios/axios/pull/5146 or will look to get my changes in https://github.com/Saghen/aw-client-js merged in before this can be merged.

ErikBjare commented:
No worries!

Just a heads-up though, there have been a few changes to master since you made this PR. You might want to carefully rebase, or resolve them some other way.

I would also appreciate it if you avoided any change to formatting rules. Big PRs like this are difficult to review as it is, so let's keep the formatting changes as minimal as possible so that the diff is readable.

ErikBjare reviewed (commented):
Oof, apparently I had all these old review comments that I'd forgotten to submit. Sorry about the late response.

ErikBjare commented:
I merged the aw-client-js PR.

Curious if you'll continue to work on this, but no pressure or rush :)

Saghen commented:
Yep! Still planning to get this wrapped up

Saghen commented:
Surprised by some of my choices on the original PR, like removing the Makefile :stuck_out_tongue: Rebased onto master but none of the changes, beside the [README change](https://github.com/ActivityWatch/aw-watcher-web/commit/b53d09997e97261e3246d1e225d12b4e4b85dd6f) seemed to be applicable. Other changes:

- added back the Makefile
- added precommit hook for formatting
- fixed up the gitignore

I should have waited for your review before squashing the commits, sorry about that. Firefox supports Manifest v3 now so I may look to upgrade to that, although there's some weird caveats wrt permissions that I might have to check on.

ErikBjare commented on Makefile:
```diff
@@ -1,50 +1,55 @@
 	# clean up
-	(cd build && rm media.zip)
-
-# Tests reproducibility of the build from srczip
-test-build-srczip: srczip build
-	(cd build/aw-watcher-web && make build)
-	@# check that aw-watcher-web.zip have the same size
-	@wc -c aw-watcher-web.zip build/aw-watcher-web/aw-watcher-web.zip | \
-		sort -n | \
-		cut -d' ' -f2 | \
-		uniq -c | \
-		grep -q ' 2 ' \
-	|| (echo "build artifacts not the same size" && exit 1)
```
We need this check still

Saghen replied:
thanks, missed this one

ErikBjare commented on .github/workflows/build.yml:
Don't reformat

ErikBjare commented on .github/workflows/codeql.yml:
Don't reformat

ErikBjare commented on static/popup.html:
It'd be a lot easier to read many of these diffs if the changes were commited as renames. Seems to have happened for `popup/style.css`, but not for the rest.

ErikBjare commented on .prettierrc:
I'd prefer to keep this config as keys in the `package.json`: https://prettier.io/docs/en/configuration.html

ErikBjare commented on package.json:
```diff
@@ -1,19 +1,31 @@
     },
-    "scripts": {
-        "build": "mkdir -p out; browserify src/*.js > out/app.js"
+    "lint-staged": {
+        "*": [
+            "npm run format"
+        ]
```
I don't see a `format` script?

Saghen commented:
Hey @ErikBjare would you be willing to take over this PR? Seems like some pretty minor things at this point and you should have edit access to the branch

Saghen commented on package.json:
```diff
@@ -1,19 +1,31 @@
+    "devDependencies": {
+        "@types/deep-equal": "^1.0.1",
+        "@types/webextension-polyfill": "^0.10.0",
+        "husky": "^9.0.11",
```
Husky is for a pre-commit hook so those aren't equivalent but I have no preference either way

ErikBjare replied:
Please remove it

ErikBjare commented:
@Saghen I haven't reviewed it properly because the diff is practically unreadable due to the formatting changes, making git not understand that files were moved and edited, not deleted and rewritten from scratch. It is possible one might be able to fix this by using rebase and adding a move commit before the other commits (as explained [here](https://stackoverflow.com/a/433114)), but I haven't done that before.

~~Given this, I don't think I could take it over in its current state without a lot of work. I would instead probably recreate the PR from scratch with inspiration taken from this, which I don't think would be faster, and which I would probably continue to procrastinate.~~ Let's see if the above thing works first, I'll give it a shot.

Saghen commented:
Gotcha, let me know how that goes! I did rewrite everything from scratch except for popup.html iirc so that's why git has not marked them as moved. And the formatting was to match the `.editorconfig` that was already in the repo but I'll take care of the review comments soon

ErikBjare commented on .editorconfig:
```diff
@@ -7,7 +7,7 @@ trim_trailing_whitespace = true
 
 # 2 spaces for JS, HTML and CSS/SASS
-[*.{js,html,css,scss,sass,less}]
+[*.{ts,js,html,css,scss,sass,less}]
```
```suggestion
[*.{ts,js,html,yml,css,scss,sass,less}]
```

ErikBjare commented on src/background/heartbeat.ts:
```diff
@@ -0,0 +1,77 @@
+    return
+  }
+
+  console.debug(`Ensuring bucket "${getBucketId()}" exists`)
+  await ensureBucket(client, getBucketId())
```
This must not run on every heartbeat.

One suggestion: If the bucket doesn't exist and the heartbeat errors, we create it.

ErikBjare commented:
> Gotcha, let me know how that goes!

It worked, but was too cumbersome to complete.

> I did rewrite everything from scratch except for popup.html and a couple others iirc so that's why git has not marked them as moved.

I realized that as I made my attempt and actually saw how significant the changes were. Even harder to review 😅, but I massively appreciate the effort of putting this together! 🥇

I'll give it another pass when the CI is passing and the formatting stuff is fixed. Might try to get another pair of eyes on this.

BelKed reviewed (changes requested):
The only concern I have is that the [Brave detection](https://github.com/ActivityWatch/aw-watcher-web/pull/126) is not present in this version of the extension, so it would be a step back in the functionality of the extension...

Everything else looks fine to me 👍

ErikBjare commented:
@BelKed Yes, we'd have to check Brave detection and other recent commits to master so that they're included correctly (as they may have disappeared in a rebase).

ErikBjare commented on src/background/helpers.ts:
```diff
@@ -0,0 +1,59 @@
+  })
+}
+
+// FIXME: Detect Brave? Using navigator.brave.isBrave does not work in a service worker
+// so this would need to be done in an injected script
+// FIXME: Detect Vivaldi? It seems to be intentionally impossible
+export const getBrowserName = () => {
+  if (
+    navigator.userAgent.includes('Opera') ||
+    navigator.userAgent.includes('OPR')
+  ) {
+    return 'opera'
+  } else if (navigator.userAgent.includes('Firefox')) {
+    return 'firefox'
+  } else if (navigator.userAgent.includes('Chrome')) {
+    return 'chrome'
+  } else if (navigator.userAgent.includes('Safari')) {
+    return 'safari'
+  } else {
+    return 'unknown'
+  }
+}
```
Only called once, so doesn't need caching like in https://github.com/ActivityWatch/aw-watcher-web/commit/0b289c4208f308050979fc729a69cf0c57dd7d8f. But that comment seems to suggest Brave is not possible with service workers? Is this a new thing in this rewrite or is it an old comment that is no longer valid?

BelKed replied:
It's probably an old comment. I've tested it now and can't see any errors when using the method from the recent PR :)

Saghen replied:
https://github.com/brave/brave-browser/issues/28101

Yep fixed upstream, just added it to this PR

Saghen commented:
@ErikBjare comments resolved and rebased on master!

BelKed commented on .github/workflows/build.yml:
```diff
@@ -1,38 +1,37 @@
+      - run: npm ci
 
+      - name: Build Firefox
+        run: npm run build:firefox
```
I think it should be `make build-firefox` as there are no scripts in the `package.json` file :) And, same for Chrome.
Maybe the `npm ci` before that isn’t necessary? (We are installing all of the packages in the `make` command)

Saghen replied:
nice catch, thank you!

ErikBjare commented:
Really excited about this, will try to find time to review and merge in the coming days.

powellnorma commented on src/background/helpers.ts:
```diff
@@ -0,0 +1,59 @@
+
+// FIXME: Detect Vivaldi? It seems to be intentionally impossible
+export const getBrowserName = () => {
+  if ((navigator as any).brave.isBrave()) {
```
TypeError: Cannot read properties of undefined (reading 'isBrave')

powellnorma commented on Makefile:
```diff
@@ -1,48 +1,85 @@
+	mkdir -p artifacts && cd build && zip ../artifacts/chrome.zip -r *
+
+zip-build-firefox:
+	mkdir -p artifacts && cd build && zip ../artifacts/firefox.zip -r *
```
Might make sense to add `-FS` to the zip command, so that not-anymore-existing files get removed in the .zip archive

ErikBjare commented:
Thank you so much for all the work you did here @Saghen, I really appreciate it!

@BelKed just dragged this across the finish line and we merged it in #152.

I applaud you for a fantastic contribution :clap: :1st_place_medal:
